feat(hogql): JSON property access

mariusandra opened the pull request:
## Problem

Deeper JSON access is one of the most requested feature for HogQL

## Changes

Implements it, but without property type autodetection --> all deeply accessed properties will be string and you need a manual `toInt()` around them.

## How did you test this code?

Wrote tests, tested in the browser:

<img width="975" alt="image" src="https://user-images.githubusercontent.com/53387/229907680-1c74f8ed-abe2-46e3-bcde-b46852d8e0f3.png">

mariusandra committed:
feat(hogql): datetime and timezones

mariusandra committed:
also for scalars and in events query

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

mariusandra committed:
a bit cleaner

mariusandra committed:
Merge branch 'hogql-timezone' of github.com:PostHog/posthog into hogql-timezone

mariusandra committed:
add tests

mariusandra committed:
uuids

github-actions[bot] committed:
Update query snapshots

mariusandra committed:
fixes

mariusandra committed:
oops

mariusandra committed:
Merge branch 'master' into hogql-timezone

mariusandra committed:
can print floats and ints now

mariusandra committed:
nans

mariusandra committed:
Merge branch 'master' into hogql-timezone

github-actions[bot] committed:
Update query snapshots

mariusandra committed:
sanitize timezone

mariusandra committed:
Merge remote-tracking branch 'origin/hogql-timezone' into hogql-timezone

mariusandra committed:
deeper json access

aluxian commented:
Awesome!

Twixes commented:
Pulled in support for `sessionStorage` in the opt-in-out classes (from #592), improved typing, and clarified some names (to e.g. `registerSession` is now `register_for_session` to align it with `register_once` and reduce conflict with the PostHog concept of a session).

thmsobrmlr commented on posthog/hogql/test/test_printer.py:
```diff
@@ -134,6 +134,29 @@ def test_hogql_properties(self):
+        try:
+            from ee.clickhouse.materialized_columns.analyze import materialize
+        except:
+            # EE not available? Assume we're good
```
Seems a bit hacky to check for EE availability like this. The test could become a noop when we move things around.

Twixes replied:
Yeah, EE tests just need to be in `ee/` rather than `posthog/`. Hard to trust a test that self-parametrizes itself this way.

thmsobrmlr reviewed (approved):
LGTM

mariusandra committed:
Update posthog/hogql/printer.py

Co-authored-by: Thomas Obermüller <thomas.obermueller@gmail.com>

mariusandra committed:
use python's ZoneInfo

github-actions[bot] committed:
Update query snapshots

mariusandra committed:
Merge branch 'hogql-timezone' of github.com:PostHog/posthog into hogql-timezone

mariusandra committed:
Merge branch 'hogql-timezone' of github.com:PostHog/posthog into hogql-timezone

Twixes reviewed (commented):
Looks awesome, just not sure about that EE-dependent test either

mariusandra committed:
split test

mariusandra committed:
move materialized import into printer

mariusandra commented:
Ugh, that test is indeed funny. I split out the materialized part into its own test, to make sure the non-materialized part runs always. This way at least we know JSON parsing works.

There's probably an even bigger bug: while `test_printer.py` was fine, `printer.py` was importing `from ee.clickhouse*` directly at the root of python file. I moved that into the function itself.

I'm not that sure I'd like to move just these materialised tests into the `ee/` folder now... 🤔

mariusandra committed:
Merge branch 'hogql-timezone' into hogql-json

mariusandra committed:
Merge branch 'master' into hogql-json

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

mariusandra committed:
remove merge bug

mariusandra committed:
Merge branch 'hogql-json' of github.com:PostHog/posthog into hogql-json

mariusandra committed:
Merge branch 'master' into hogql-json

github-actions[bot] committed:
Update query snapshots
